### ISSUE TITLE
Zenject is too slow in large scenes

### ISSUE THREAD

**Issue by HDANILO** [ssannandeji/Zenject-2019#188]
In the project we're working at Lumentech, our scene has A LOT of components, before I started to use the Zenject, our scene loading was something like 20s, after we started to use it in this project it went to 120s, both times measured using an IPAD 3. I put a benchmark inside the SceneContext Awake and discovered the reason it was wasting so much more time, the zenject was trying to inject and resolve dependencies of EVERY COMPONENT IN THE SCENE.

Ok, I get it, it is a way to do, since this is a generic library you never know when somebody is trying to do an inheritance using transform and need an injection there, but the increase in time was unacceptable.

I solved the problem putting this naughty if inside the `GetInjectableComponents ` of ContextUtil.cs

> if (component is Transform || 
>      component is RectTransform || 
>      component is CanvasRenderer || 
>      component is ILayoutElement || 
>      component is TextMeshProUGUI ||
>      component is CanvasGroup || 
>      component is Button || 
>      component is ContentSizeFitter || 
>      component is Collider || 
>      component is IsoObject || 
>      component is Renderer) 
> {
>      continue;
> }

Turns out our scene had 17k transforms and 10k more others components and the zenject was trying to inject in every one of them. That list sums up the bigger amount of components that didn't use injection but Zenject was trying to do anyway.

One solution to the problem is to create an ScriptableObject where developers can add types they want to ignore

Just to clarify, the benchmarks that I ran was using the real game that we're working on, not a test scene, we really have that much things in the scene. The game is already in soft launch and runs quite well on medium to low devices.

**Comment by svermeulen** [comment]
Actually, you make a great point... It should only inject into MonoBehaviours, not Component.  As far as I know, it's not possible to derive from Component directly except for built in Unity classes, so we should be safe to assume that there isn't any injected data except for MonoBehaviour derived classes.  That would address a bunch of the types you included there in your if statement.

We could also add a ScriptableObject to contain types to filter out like you say.  Or maybe we could pre-calculate a whitelist that contains all types that have injectables at edit time.  That would also avoid having Zenject pointlessly use costly reflection for lots of types that have no injectables.   I would like to understand more about exactly what part of the code is causing the slowdown first though, for cases like this when you have thousands of components in the scene.  Like is it reflection or is it just calling `GetComponents<>` on every game object in the scene?  The fact that adding your if statement addressed the issue for you seems to suggest that it wasn't the latter.

**Comment by HDANILO** [comment]
I forgot to mention that after filtering that out, the SceneContext went to 500ms. I'm pretty sure the problem is the reflection. If the assembly type table grows or If the class has a lot of members the reflection also can grow even in time consumption.

I don't think we'll manage to get rid of the reflection, we can try to optimize it, but since this little change brought 120s to 500ms, we should start doing the type filtering

**Comment by Jamjardavies** [comment]
Could we have an InjectMonoBehaviour, and anything you want auto-injected extend this instead of MonoBehaviour? That will remove all scripts and only do injection on ones we state?

**Comment by HDANILO** [comment]
Also, another thing about instantiating prefabs using Zenject: it first instantiate the prefab and then it reparent If its necessary. Since unity 5.4 reparenting is very expensive, its a good Idea to instantiate directly as a child of the parent.

Sorry for any mispelling Im writing using my cellphone that is trying to autocorrect to another language(portuguese)

**Comment by Jamjardavies** [comment]
Oh interesting, did not know this!

**Comment by HDANILO** [comment]
Unity keeps a tree like buffer instead of set(it was set before). Inserting a new element in the middle of the tree forces the buffer to push every element forth causing a recusive copy of each element after the one you're inserting in the tree. There is one Unite 2016 vídeo about optimization that they talk about it

**Comment by HDANILO** [comment]
Ok, I thought about it a little bit I've come with a list of good and bad things for both approachs:

The MonoBehaviourInject approach is a cleaner one, it is transparent to the developer that it'll be injected, BUT, more components to game objects, more serializable data to be deserialized once the game/scene loads. Also, it'll be like hell to manually manage and add MonoBehaviourInject in each GameObject. A solution to this would be use the RequireComponent attribute, but if we do, I don't think we can destroy it afterwards.

Pros:
* Transparent process

Cons:
* Hell to manage

About pre-computing types that have inject, I prefer this solution, but it needs to be done with carefully since it can make the editor very burden. I don't know if unity has a pre-compile function but if it does, we can cache some hash to skip processing all injection types, making the process very simple, easy to manage and even fast.

Either way, it would be good if we can avoid those IEnumerable types, it can generate a lot of garbage and cause stuttering. I dont feel like its high priority, because that really is an issue on the Unity's side. Perhaps they feel ashamed and upgrade that old garbage collector. I love IEnumerables.

**Comment by bvance** [comment]
Lately I've been toying with the notion of pre-computing + serializing not only a list of all types that require injection, but precisely which of their members need to be injected. I wonder if it would be very much faster?

**Comment by Jamjardavies** [comment]
@HDANILO When I say InjectMonoBehaviour, I don't mean a script you add to an object to allow it, but instead, change your scripts to inherit that one, instead of the standard MonoBehaviour. You wouldn't have to do any changes in the editor at all, just the few classes you'd like to inject.

Regarding putting the created GameObject into the child directly, I'll look to see if I can change this and do a PR for Steve to look at.

@bvance Interesting idea, baking part of the reflection so you don't need to do it at runtime?

**Comment by HDANILO** [comment]
@bvance I believe protobuf does something like that and it increases the serialization/deserialization performance

**Comment by HDANILO** [comment]
I was thinking, we all know that reflection is too bad in mobile devices, so I thought an way to work a round it.
It would be bad to create a Injectable<T> type to be used instead of the [Inject] attribute?

There would be no need to use reflection and the evaluation would be lazy, it would resolve only when the proper value attribute was called, and could be wrap around like the nullable is. I wouldn't be bothered to use it and avoid some reflection, I'm even considering using Container.Resolve<T> for classes that I have many instances in order to reduce the reflections calls.

Ofc we should keep [Inject] too as it can be used on some use cases like fetching an instance with a specific id.

What are you guys thoughts on that?

**Comment by Jamjardavies** [comment]
The problem with lazy instantiation etc is memory fragmentation.

As a rule of thumb, I try and create everything up-front and only do mid-game instantiation for things I know I'll be destroying and cleaning up with GC.

There is a new Lazy<T> in Zenject, although it's undocumented I believe. This does this behaviour but I think there is still a reflection overhead.

I've not had much problems with the reflection calls in the past, on both mobile and console, but I do agree, it could be faster. Faster loading is always better.

**Comment by svermeulen** [comment]
Yeah, support for `Lazy<>` is in the next update (4.8) which I'm pushing out soon.  I don't think that will be of any help with this issue though.  It might actually be worse, because it won't perform the reflection until the value is actually used, which could be after startup in the middle of rendering.

Anyway, I do really like this idea of avoiding reflection altogether for classes that you know don't have any injectables.  That could be a huge win.  So I'd like to take a look at something here for 4.9.  Unity does have an editor callback for when all the code is compiled, so we could automatically update a file that stores some pre-calculated information we could use.

@Jamjardavies I also didn't know that instantiating directly on a parent was less expensive than instantiating, then parenting.  If you want to take a look at that, feel free.  It might not be too bad to just add an optional parameter to the instantiate methods

**Comment by HDANILO** [comment]
@svermeulen the `Resolve<T>` method does use reflection? because if it doesnt, the `Lazy<T>` should use it

**Comment by svermeulen** [comment]
Yes, resolve will also trigger reflection (but only if its the first time the type has been instantiated).  There's no way around it when instantiating via zenject, unless we do something like @bvance suggests with serializing out the reflection data, but even in that case you still have to use reflection to get the actual injectable fields/properties/methods.  That could still offer some benefits though, would need to investigate

**Comment by HDANILO** [comment]
Oh I see, you're talking about the activator reflection, I was talking/thinking about the method/attribute finding reflection. But yeah, I think Lazy<T> does what I thought it would

**Comment by Jamjardavies** [comment]
@svermeulen I'll have a look at that tonight and do a PR on develop :)

**Comment by Jamjardavies** [comment]
I've been investigating a way to potentially remove most, if not all reflection inside Zenject.

So far, tests are proving promising, but I need to test more, however, would be willing to talk about it in more detail, if you're free and online.

**Comment by HDANILO** [comment]
I'm online if you want to talk about it

**Comment by svermeulen** [comment]
I looked into this issue a bit more.  I created a scene with 30k transforms like in the original description here and was able to reproduce the slowdown.  As it turned out, the reflection wasn't actually the biggest culprit (though that remains a separate issue that we need to address).  Instead, I actually found that just the search over all the transforms was a bigger performance hit.  So I took a look at the method doing that and found a different approach to it that was much faster (in my test scene, the new approach reduced the startup time for SceneContext down from 600ms to 40ms).  You can try it on develop branch if interested

I'd also like to see what we can do about the reflection related performance hits so I'll leave this item open for that

**Comment by greenecreations** [comment]
I see some commits from @yonkahlon / @svermeulen, so I was wondering when this performance enhancement would be available for everyone to use?

I'm new to unity3d but I've been a professional AngularJS dev for years, so the first thing I looked up was "where is Unity3d's DI." I'd love to use Zenject, but I can't lose any performance, you know?

P.S. thank you to everyone that has worked on Zenject. I hope to use it so I can evangelize it to friends and to the local game development community in Philadelphia.

**Comment by svermeulen** [comment]
I can sympathize with your being unable to sacrifice any performance.  The performance hit given by the reflection is a deterrent for other people too.  I like the ideas in this thread to address it though, and would like to give them a shot soon, but just haven't had the time yet, and not sure when that will change

**Comment by saynomoo** [comment]
I was investigating the Zenject performance from our perspective a bit. I tried a little micro benchmark with about 100 000 game object in the hierachy. The startup time of the scene without zenject  was about 7s ("injecting" done in Awake()). When turning on Zenject, the startup time of the scene wen't all the way to 85s.

It turned out that in this case the main culprit was LazyInstanceInjector.LazyInjectAll. I modified the method so that the objects are not removed from the HashSet one by one, but replacing the whole hashset at once. This took the scene startup time back down to 11s.

In our real world scene where there are close to 10 000 gameobjects (which starts in about 10s), the difference was, however, unnoticeable. So it did not help us, but it might help in some bigger scenes.

**Comment by saynomoo** [comment]
Btw, a nice way to get rid of the reflection based Injection could be an Editor-script that creates a partial class for the injectable class. That class implements an interface with method inject, which would then set the references "manually".

**Comment by saynomoo** [comment]
Oh, and getting rid of allocations in ZenUtilInternal.GetInjectableMonoBehaviours improved the startup time in the micro benchmark from 11s to 8.5s. I implemented the method without the sorting and toLists directly with yields.

**Comment by svermeulen** [comment]
@saynomoo Sounds really helpful.  Any chance you'd be willing to do a PR or failing that share some of the code?

**Comment by saynomoo** [comment]
@svermeulen : The code I used was for benchmarking purposes, and unfortunately the functionality was not 100% same. So I wouldn't take them as is to Zenject. For example the LazyInstanceInjector modification is as follows: 

```
        public void LazyInjectAll()
        {
            var toInject = _instancesToInject;
            _instancesToInject = new HashSet<object>();
            foreach (var instance in toInject)
            {
                _container.Inject(instance);
            }
        }
```

The difference to the existing code is that it fails to inject instances that are added to the injector during the injection phase.

The GetInjectableMonobehaviours, on the other hand, skips the sorting, which is not stricly backwards compatible:
```
	protected override IEnumerable<MonoBehaviour> GetInjectableMonoBehaviours()
	{
		foreach (var go in gameObject.scene.GetRootGameObjects())
		{
			if (go != null)
			{
				foreach (var childMb in GetInjectableMonoBehaviours(go))
				{
					yield return childMb;
				}
			}

		}
	}

	protected IEnumerable<MonoBehaviour> GetInjectableMonoBehaviours(GameObject go)
	{
		var doNest = true;
		foreach (var monoBehaviour in go.GetComponents<MonoBehaviour>())
		{
			if (monoBehaviour == null || monoBehaviour == this || typeof(MonoInstaller).IsAssignableFrom(monoBehaviour.GetType()))
			{
				continue;
			}
			if (typeof(GameObjectContext).IsAssignableFrom(monoBehaviour.GetType()))
			{
				doNest = false;
			}

			yield return monoBehaviour;
		}

		if (!doNest) yield break;
		for (var i = go.transform.childCount - 1; i >= 0; i--)
		{
			var child = go.transform.GetChild(i);
			if (child == null) continue;
			foreach (var childMb in GetInjectableMonoBehaviours(child.gameObject))
			{
				yield return childMb;
			}
		}
	}

```

**Comment by svermeulen** [comment]
Ok I committed a similar change to the LazyInjectAll method and noticed huge improvements with scenes with 100k+ transforms.  In my test scene, that simple change to that method brought the startup time down from 80 seconds to 6.  It seems that calling HashSet.First() can sometimes be a performance heavy operation.  Thanks.

**Comment by svermeulen** [comment]
Added a few changes relevant to this issue:
- Added those optimizations to the GetInjectableMonoBehaviours suggested by @saynomoo.  This did speed up SceneContext.Awake a fair amount actually.  In my test (with 100k transforms) it went from 0.82 seconds to to 0.4 seconds.
- Changed to avoid doing reflection on any class from the UnityEngine namespace
- Other minor memory/performance allocations

Still more optimizations to do from this thread but some progress at least.

The Inject / Bind / Resolve / Instantiate operations still have a bunch of memory allocs in them, but these operations should only occur at initialization time so to me it doesn't seem like a big deal right now.  For dynamic objects that do these operations (eg. GameObjectContext) you can use memory pools to execute all allocs at initialization time as well.  So Zenject shouldn't prevent people from maintaining zero per-frame allocs.

**Comment by svermeulen** [comment]
Ok I added what I think is the last remaining issue in this thread:  build-time reflection baking.  It is working in source (and documented) and will be included in version 6.5.0 and 7.3.0
